File: Backtracking/DFS/79WordSearch.py
from typing import List

# 经典DFS, 只不过在查看children结果的时候, 要用OR, 只要有一条路径是True, 就返回True
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        dirs = [[-1, 0], [1, 0], [0, 1], [0, -1]]
        visited = set()
        def dfs(row, col, index):
            if index == len(word): return True
            
            if 0 <= row < len(board) and 0 <= col < len(board[0]):
                if board[row][col] == word[index] and ((row, col) not in visited):
                    visited.add((row, col))
                    res = False
                    for [i, j] in dirs:
                        res = res or dfs(row+i, col+j, index + 1)
                    visited.remove((row, col))
                    return res
            else: return False
        
        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] == word[0]:
                    if bfs(i, j, 0):
                        return True
        return False

File: Backtracking/DFS/List/79WordSearch.java
package List;

class Solution {
    public boolean exist(char[][] board, String word) {
        var numRow = board.length;
        var numCol = board[0].length;
        boolean[][] visited = new boolean[numRow][numCol];
        boolean result = false;

        for (int i = 0; i < numRow; i++) {
            for (int j = 0; j < numCol; j++) {
                if (board[i][j] == word.charAt(0)) {
                    result = dfs(i, j, board, word, 0, visited);
                    if (result)
                        return true;
                }
            }
        }
        return false;
    }

    public boolean dfs(int i, int j, char[][] board, String word, int index, boolean[][] visited) {
        if (index == word.length()) {
            return true;
        }

        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || visited[i][j] == true
                || board[i][j] != word.charAt(index)) {
            return false;
        }
        visited[i][j] = true;
        if (dfs(i + 1, j, board, word, index + 1, visited) ||
                dfs(i, j + 1, board, word, index + 1, visited) ||
                dfs(i - 1, j, board, word, index + 1, visited) ||
                dfs(i, j - 1, board, word, index + 1, visited)) {
            return true;
        }
        visited[i][j] = false;
        return false;
    }
}